src/operator/join.rs

```diff
@@ -0,0 +1,215 @@
+    for<'a> &'a I1: IntoIterator<Item = (&'a K, &'a I1::ZSet)>,
+    for<'a> &'a I1::ZSet: IntoIterator<Item = (&'a V1, &'a W)>,
+    I2: IndexedZSet<K, V2, W>,
+    for<'a> &'a I2::ZSet: IntoIterator<Item = (&'a V2, &'a W)>,
+    F: Fn(&K, &V1, &V2) -> V,
+    Z: MapBuilder<V, W> + 'static,
+{
+    let mut map = Z::empty();
+    for (k, vals1) in i1.into_iter() {
+        if let Some(vals2) = i2.get_in_support(k) {
+            for (v1, w1) in vals1.into_iter() {
+                for (v2, w2) in vals2.into_iter() {
+                    map.increment_owned(join_func(k, v1, v2), w1.mul_by_ref(w2));
+                }
+            }
+        }
+    }
+    map
+}
+
+impl<K, V1, V2, V, W, F, I1, I2, Z> BinaryOperator<I1, I2, Z>
+    for Join<K, V1, V2, V, W, F, I1, I2, Z>
+where
+    K: KeyProperties,
+    V1: KeyProperties,
+    V2: KeyProperties,
+    W: ZRingValue,
+    I1: IndexedZSet<K, V1, W>,
+    for<'a> &'a I1: IntoIterator<Item = (&'a K, &'a I1::ZSet)>,
+    for<'a> &'a I1::ZSet: IntoIterator<Item = (&'a V1, &'a W)>,
+    I2: IndexedZSet<K, V2, W>,
+    for<'a> &'a I2: IntoIterator<Item = (&'a K, &'a I2::ZSet)>,
+    for<'a> &'a I2::ZSet: IntoIterator<Item = (&'a V2, &'a W)>,
+    F: Fn(&K, &V1, &V2) -> V + 'static,
+    V: 'static,
+    Z: MapBuilder<V, W> + 'static,
+{
+    fn eval(&mut self, i1: &I1, i2: &I2) -> Z {
+        // Iterate over smaller index.
+        if i1.support_size() < i2.support_size() {
+            join_inner(i1, i2, &self.join_func)
+        } else {
+            join_inner(i2, i1, &|k, i2, i1| (self.join_func)(k, i1, i2))
+        }
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use crate::{
+        algebra::FiniteHashMap,
+        circuit::{Root, Stream},
+        finite_map,
+        operator::Generator,
+    };
+
+    #[test]
+    fn join_test() {
+        let root = Root::build(move |circuit| {
+            let mut input1 = vec![vec![
+                ((1, "a"), 1),
+                ((1, "b"), 2),
+                ((2, "c"), 3),
+                ((2, "d"), 4),
+                ((3, "e"), 5),
+                ((3, "f"), -2),
+            ]]
+            .into_iter();
+            let mut input2 = vec![vec![
+                ((2, "g"), 3),
+                ((2, "h"), 4),
+                ((3, "i"), 5),
+                ((3, "j"), -2),
+                ((4, "k"), 5),
+                ((4, "l"), -2),
+            ]]
+            .into_iter();
+            let mut outputs = vec![finite_map! {
+                (2, "c g".to_string()) => 9,
+                (2, "c h".to_string()) => 12,
+                (2, "d g".to_string()) => 12,
+                (2, "d h".to_string()) => 16,
+                (3, "e i".to_string()) => 25,
+                (3, "e j".to_string()) => -10,
+                (3, "f i".to_string()) => -10,
+                (3, "f j".to_string()) => 4
+            }]
+            .into_iter();
+            let index1: Stream<_, FiniteHashMap<usize, FiniteHashMap<&'static str, isize>>> =
```

Comment:
it's a bummer that you have to specify the type for index1.
How come DD doesn't?

Reply:
Because DD has exactly one type to represent Z-Sets (`Collection`) and one type to represent indexed Z-sets (`Arrangement`). We can have a convenience API on top of this that avoids type annotations by using fixed types.